### ISSUE TITLE
test_deepseek_v3.py is much slower using nvFuser than using torch-tp

### ISSUE THREAD

**Issue by wujingyue** [NVIDIA/Fuser#4627]
```
me @ viking-prod-232 : dev | /opt/pytorch/nvfuser (wjy/benchmark)
$ mpirun -np 8 -output-filename /tmp/test_deepseek_v3 pytest tests/python/multidevice/test_deepseek_v3.py --only-mpi

# Other ranks show similar results.
$ cat /tmp/test_deepseek_v3/1/rank.0/stdout

----------------------------------------------------------------------------------------- benchmark: 2 tests -----------------------------------------------------------------------------------------
Name (time in ms)                        Min                Max               Mean             StdDev             Median                IQR            Outliers      OPS            Rounds  Iterations
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_transformer_layer[TORCH_TP]     24.4076 (1.0)      29.4599 (1.0)      26.2417 (1.0)       2.0156 (1.0)      25.6295 (1.0)       2.7083 (1.0)           1;0  38.1073 (1.0)           5           1
test_transformer_layer[NVFUSER]      28.6499 (1.17)     75.5864 (2.57)     39.0602 (1.49)     20.4520 (10.15)    29.8216 (1.16)     13.5153 (4.99)          1;1  25.6015 (0.67)          5           1
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```

originally posted in https://github.com/NVIDIA/Fuser/pull/4601

The biggest difference I've found so far is the time spent on ncclAllReduce: https://github.com/NVIDIA/Fuser/pull/4601#issuecomment-2957198101. The first half of allreduces are 10x slower with nvFuser than with torch. Note the second column is the duration in nanoseconds. 

```
me @ viking-prod-231 : dev | /opt/pytorch/nvfuser (8dfe66cf)
$ nsys stats --report=cuda_gpu_trace ~/workspace/nvfuser.nsys-rep | grep -i allreduce
  206424976         766205    9164      24     1     1   640     1     1       96         0.022         0.082                                                     NVIDIA H100 80GB HBM3 (0)    1              24  ncclDevKernel_AllReduce_Sum_bf16_RING_LL(ncclDevKernelArgsStorage<(unsigned long)4096>)
  209832450        1667705    9838      24     1     1   640     1     1       96         0.022         0.082                                                     NVIDIA H100 80GB HBM3 (0)    1              24  ncclDevKernel_AllReduce_Sum_bf16_RING_LL(ncclDevKernelArgsStorage<(unsigned long)4096>)
  211565306        1796537    9972      24     1     1   640     1     1       96         0.022         0.082                                                     NVIDIA H100 80GB HBM3 (0)    1              24  ncclDevKernel_AllReduce_Sum_bf16_RING_LL(ncclDevKernelArgsStorage<(unsigned long)4096>)
  213423250        1155708   10100      24     1     1   640     1     1       96         0.022         0.082                                                     NVIDIA H100 80GB HBM3 (0)    1              24  ncclDevKernel_AllReduce_Sum_bf16_RING_LL(ncclDevKernelArgsStorage<(unsigned long)4096>)
  214644269        1146683   10235      24     1     1   640     1     1       96         0.022         0.082                                                     NVIDIA H100 80GB HBM3 (0)    1              24  ncclDevKernel_AllReduce_Sum_bf16_RING_LL(ncclDevKernelArgsStorage<(unsigned long)4096>)
  215856072        1110075   10364      24     1     1   640     1     1       96         0.022         0.082                                                     NVIDIA H100 80GB HBM3 (0)    1              24  ncclDevKernel_AllReduce_Sum_bf16_RING_LL(ncclDevKernelArgsStorage<(unsigned long)4096>)
  217030499        1145851   10493      24     1     1   640     1     1       96         0.022         0.082                                                     NVIDIA H100 80GB HBM3 (0)    1              24  ncclDevKernel_AllReduce_Sum_bf16_RING_LL(ncclDevKernelArgsStorage<(unsigned long)4096>)
  218241150        1075771   10622      24     1     1   640     1     1       96         0.022         0.082                                                     NVIDIA H100 80GB HBM3 (0)    1              24  ncclDevKernel_AllReduce_Sum_bf16_RING_LL(ncclDevKernelArgsStorage<(unsigned long)4096>)
  219645112        1445594   10754      24     1     1   640     1     1       96         0.022         0.082                                                     NVIDIA H100 80GB HBM3 (0)    1              24  ncclDevKernel_AllReduce_Sum_bf16_RING_LL(ncclDevKernelArgsStorage<(unsigned long)4096>)
  221153874        1194074   10882      24     1     1   640     1     1       96         0.022         0.082                                                     NVIDIA H100 80GB HBM3 (0)    1              24  ncclDevKernel_AllReduce_Sum_bf16_RING_LL(ncclDevKernelArgsStorage<(unsigned long)4096>)
  222406348        1124987   11010      24     1     1   640     1     1       96         0.022         0.082                                                     NVIDIA H100 80GB HBM3 (0)    1              24  ncclDevKernel_AllReduce_Sum_bf16_RING_LL(ncclDevKernelArgsStorage<(unsigned long)4096>)
  223798950         888445   11138      24     1     1   640     1     1       96         0.022         0.082                                                     NVIDIA H100 80GB HBM3 (0)    1              24  ncclDevKernel_AllReduce_Sum_bf16_RING_LL(ncclDevKernelArgsStorage<(unsigned long)4096>)
  226346811         388991   11275      24     1     1   640     1     1       96         0.022         0.082                                                     NVIDIA H100 80GB HBM3 (0)    1              24  ncclDevKernel_AllReduce_Sum_bf16_RING_LL(ncclDevKernelArgsStorage<(unsigned long)4096>)
  227936277         190143   11409      24     1     1   640     1     1       96         0.022         0.082                                                     NVIDIA H100 80GB HBM3 (0)    1              24  ncclDevKernel_AllReduce_Sum_bf16_RING_LL(ncclDevKernelArgsStorage<(unsigned long)4096>)
  229213039          67296   11543      24     1     1   512     1     1       96         0.022         0.082                                                     NVIDIA H100 80GB HBM3 (0)    1              24  ncclDevKernel_AllReduce_Sum_bf16_RING_LL(ncclDevKernelArgsStorage<(unsigned long)4096>)
  230568105          23392   11677      21     1     1   512     1     1       96         0.022         0.082                                                     NVIDIA H100 80GB HBM3 (0)    1              24  ncclDevKernel_AllReduce_Sum_bf16_RING_LL(ncclDevKernelArgsStorage<(unsigned long)4096>)
  231853796          23104   11811       3     1     1   512     1     1       96         0.022         0.082                                                     NVIDIA H100 80GB HBM3 (0)    1              24  ncclDevKernel_AllReduce_Sum_bf16_RING_LL(ncclDevKernelArgsStorage<(unsigned long)4096>)
  233127103          17888   11945       1     1     1   128     1     1       96         0.022         0.082                                                     NVIDIA H100 80GB HBM3 (0)    1              24  ncclDevKernel_AllReduce_Sum_bf16_RING_LL(ncclDevKernelArgsStorage<(unsigned long)4096>)
  234459769         182975   12074      24     1     1   640     1     1       96         0.022         0.082                                                     NVIDIA H100 80GB HBM3 (0)    1              24  ncclDevKernel_AllReduce_Sum_bf16_RING_LL(ncclDevKernelArgsStorage<(unsigned long)4096>)
  235739283         185440   12203      24     1     1   640     1     1       96         0.022         0.082                                                     NVIDIA H100 80GB HBM3 (0)    1              24  ncclDevKernel_AllReduce_Sum_bf16_RING_LL(ncclDevKernelArgsStorage<(unsigned long)4096>)
  237915498         189343   12407      24     1     1   640     1     1       96         0.022         0.082                                                     NVIDIA H100 80GB HBM3 (0)    1              24  ncclDevKernel_AllReduce_Sum_bf16_RING_LL(ncclDevKernelArgsStorage<(unsigned long)4096>)
me @ viking-prod-231 : dev | /opt/pytorch/nvfuser (8dfe66cf)
$ nsys stats --report=cuda_gpu_trace ~/workspace/torch.nsys-rep | grep -i allreduce
  163692376         754077    7806      24     1     1   640     1     1       96         0.022         0.082                                                     NVIDIA H100 80GB HBM3 (0)    1              20  ncclDevKernel_AllReduce_Sum_bf16_RING_LL(ncclDevKernelArgsStorage<(unsigned long)4096>)
  171608535          92928    8481      24     1     1   640     1     1       96         0.022         0.082                                                     NVIDIA H100 80GB HBM3 (0)    1              20  ncclDevKernel_AllReduce_Sum_bf16_RING_LL(ncclDevKernelArgsStorage<(unsigned long)4096>)
  175501639          94624    8605      24     1     1   640     1     1       96         0.022         0.082                                                     NVIDIA H100 80GB HBM3 (0)    1              20  ncclDevKernel_AllReduce_Sum_bf16_RING_LL(ncclDevKernelArgsStorage<(unsigned long)4096>)
  182863785          89472    8723      24     1     1   640     1     1       96         0.022         0.082                                                     NVIDIA H100 80GB HBM3 (0)    1              20  ncclDevKernel_AllReduce_Sum_bf16_RING_LL(ncclDevKernelArgsStorage<(unsigned long)4096>)
  186687066          92095    8848      24     1     1   640     1     1       96         0.022         0.082                                                     NVIDIA H100 80GB HBM3 (0)    1              20  ncclDevKernel_AllReduce_Sum_bf16_RING_LL(ncclDevKernelArgsStorage<(unsigned long)4096>)
  190563818          92640    8967      24     1     1   640     1     1       96         0.022         0.082                                                     NVIDIA H100 80GB HBM3 (0)    1              20  ncclDevKernel_AllReduce_Sum_bf16_RING_LL(ncclDevKernelArgsStorage<(unsigned long)4096>)
  198051820          90463    9086      24     1     1   640     1     1       96         0.022         0.082                                                     NVIDIA H100 80GB HBM3 (0)    1              20  ncclDevKernel_AllReduce_Sum_bf16_RING_LL(ncclDevKernelArgsStorage<(unsigned long)4096>)
  205560237          90528    9205      24     1     1   640     1     1       96         0.022         0.082                                                     NVIDIA H100 80GB HBM3 (0)    1              20  ncclDevKernel_AllReduce_Sum_bf16_RING_LL(ncclDevKernelArgsStorage<(unsigned long)4096>)
  209340445          89984    9323      24     1     1   640     1     1       96         0.022         0.082                                                     NVIDIA H100 80GB HBM3 (0)    1              20  ncclDevKernel_AllReduce_Sum_bf16_RING_LL(ncclDevKernelArgsStorage<(unsigned long)4096>)
  216793183          88992    9441      24     1     1   640     1     1       96         0.022         0.082                                                     NVIDIA H100 80GB HBM3 (0)    1              20  ncclDevKernel_AllReduce_Sum_bf16_RING_LL(ncclDevKernelArgsStorage<(unsigned long)4096>)
  224286848          87264    9559      24     1     1   640     1     1       96         0.022         0.082                                                     NVIDIA H100 80GB HBM3 (0)    1              20  ncclDevKernel_AllReduce_Sum_bf16_RING_LL(ncclDevKernelArgsStorage<(unsigned long)4096>)
  228048881          89472    9677      24     1     1   640     1     1       96         0.022         0.082                                                     NVIDIA H100 80GB HBM3 (0)    1              20  ncclDevKernel_AllReduce_Sum_bf16_RING_LL(ncclDevKernelArgsStorage<(unsigned long)4096>)
  235479891         184575    9800      24     1     1   640     1     1       96         0.022         0.082                                                     NVIDIA H100 80GB HBM3 (0)    1              20  ncclDevKernel_AllReduce_Sum_bf16_RING_LL(ncclDevKernelArgsStorage<(unsigned long)4096>)
  242813493          61600    9924      24     1     1   640     1     1       96         0.022         0.082                                                     NVIDIA H100 80GB HBM3 (0)    1              20  ncclDevKernel_AllReduce_Sum_bf16_RING_LL(ncclDevKernelArgsStorage<(unsigned long)4096>)
  246541510          38656   10048      24     1     1   512     1     1       96         0.022         0.082                                                     NVIDIA H100 80GB HBM3 (0)    1              20  ncclDevKernel_AllReduce_Sum_bf16_RING_LL(ncclDevKernelArgsStorage<(unsigned long)4096>)
  253813736          23936   10174      21     1     1   512     1     1       96         0.022         0.082                                                     NVIDIA H100 80GB HBM3 (0)    1              20  ncclDevKernel_AllReduce_Sum_bf16_RING_LL(ncclDevKernelArgsStorage<(unsigned long)4096>)
  257529241          23712   10401       3     1     1   512     1     1       96         0.022         0.082                                                     NVIDIA H100 80GB HBM3 (0)    1              20  ncclDevKernel_AllReduce_Sum_bf16_RING_LL(ncclDevKernelArgsStorage<(unsigned long)4096>)
  261234506          20352   10525       1     1     1   128     1     1       96         0.022         0.082                                                     NVIDIA H100 80GB HBM3 (0)    1              20  ncclDevKernel_AllReduce_Sum_bf16_RING_LL(ncclDevKernelArgsStorage<(unsigned long)4096>)
  264906395         185855   10648      24     1     1   640     1     1       96         0.022         0.082                                                     NVIDIA H100 80GB HBM3 (0)    1              20  ncclDevKernel_AllReduce_Sum_bf16_RING_LL(ncclDevKernelArgsStorage<(unsigned long)4096>)
  272024926         185503   10767      24     1     1   640     1     1       96         0.022         0.082                                                     NVIDIA H100 80GB HBM3 (0)    1              20  ncclDevKernel_AllReduce_Sum_bf16_RING_LL(ncclDevKernelArgsStorage<(unsigned long)4096>)
  275663599         186687   11170      24     1     1   640     1     1       96         0.022         0.082                                                     NVIDIA H100 80GB HBM3 (0)    1              20  ncclDevKernel_AllReduce_Sum_bf16_RING_LL(ncclDevKernelArgsStorage<(unsigned long)4096>)
```

**Comment by wujingyue** [comment]
I don't find any meaningful differences in the NCCL trace. Both runs call the same NCCL kernel (`ncclDevKernel_AllReduce_Sum_bf16_RING_LL(ncclDevKernelArgsStorage<(unsigned long)4096>)
`) with the same size, the same launch parameters, and 512-byte aligned addresses. All allreduces are in-place. The traces even show exactly the same simulated time. 

Details:

With @nsarka's help, I was able to dump `NCCL_DEBUG=TRACE NCCL_DEBUG_SUBSYS=COLL`. This is done by
```
$ pwd
/home/me
$ git clone https://github.com/NVIDIA/nccl.git
$ cd nccl
$ make TRACE=1 -j src.build
```
and then running `mpirun` with `-x LD_PRELOAD=/home/me/nccl/build/lib/libnccl.so.2.27.3 -x NCCL_DEBUG=TRACE -x NCCL_DEBUG_SUBSYS=COLL`. 

[nvfuser_trace.txt](https://github.com/user-attachments/files/20714347/nvfuser_trace.txt)

[torch_trace.txt](https://github.com/user-attachments/files/20714349/torch_trace.txt)

**Comment by wujingyue** [comment]
The diff is likely due to host latency rather than ncclAllReduce. 

I added the following code to time the expert loop in the HuggingFace model. 

```diff
--- /home/me/.cache/huggingface/modules/transformers_modules/deepseek-ai/deepseek-v3/e815299b0bcbac849fa540c768ef21845365c9eb/modeling_deepseek.py      2025-06-12 22:36:55.345522770 -0700
+++ /tmp/modeling_deepseek.py   2025-06-12 22:36:23.280063694 -0700
@@ -573,6 +573,8 @@
             tokens_per_expert = tokens_per_expert_post_gather
         tokens_per_expert = tokens_per_expert.cpu().numpy()

+        import time
+        start = time.time()
         outputs = []
         start_idx = 0
         for i, num_tokens in enumerate(tokens_per_expert):
@@ -584,6 +586,8 @@
             expert_out = expert(tokens_for_this_expert)
             outputs.append(expert_out)
             start_idx = end_idx
+        end = time.time()
+        print(f"Routed expert loop: {end - start:.6f} seconds")

         outs = torch.cat(outputs, dim=0) if len(outputs) else sorted_tokens.new_empty(0)
         if self.ep_size > 1:
```

```
$ mpirun -np 2 pytest tests/python/multidevice/test_deepseek_v3.py --only-mpi -s
Routed expert loop: 0.136883 seconds
Routed expert loop: 0.140009 seconds
Routed expert loop: 0.016674 seconds
Routed expert loop: 0.017179 seconds
Routed expert loop: 0.016736 seconds
Routed expert loop: 0.017200 seconds
Routed expert loop: 0.016377 seconds
Routed expert loop: 0.016449 seconds
Routed expert loop: 0.016165 seconds
Routed expert loop: 0.016478 seconds
Routed expert loop: 0.016257 seconds
Routed expert loop: 0.016586 seconds
..Routed expert loop: 0.034522 seconds
Routed expert loop: 0.034957 seconds
Routed expert loop: 0.020913 seconds
Routed expert loop: 0.021353 seconds
Routed expert loop: 0.023029 seconds
Routed expert loop: 0.023065 seconds
Routed expert loop: 0.020878 seconds
Routed expert loop: 0.021034 seconds
Routed expert loop: 0.020512 seconds
Routed expert loop: 0.021004 seconds
Routed expert loop: 0.020521 seconds
Routed expert loop: 0.021005 seconds
..
----------------------------------------------------------------------------------------- benchmark: 2 tests -----------------------------------------------------------------------------------------
Name (time in ms)                        Min                Max               Mean             StdDev             Median                IQR            Outliers      OPS            Rounds  Iterations
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_transformer_layer[TORCH_TP]     25.8431 (1.0)      57.4801 (1.0)      32.7181 (1.0)      13.8849 (1.0)      25.9604 (1.0)       9.7894 (1.0)           1;1  30.5641 (1.0)           5           1
test_transformer_layer[NVFUSER]      30.3425 (1.17)     92.1303 (1.60)     43.2442 (1.32)     27.3409 (1.97)     31.0179 (1.19)     16.9748 (1.73)          1;1  23.1245 (0.76)          5           1
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```

The first ten prints are from TORCH_TP and the rest are from NVFUSER. Ignoring the warmup run, NVFUSER was consistently slower than TORCH_TP by ~5 milliseconds. This diff is aligned with the wall time difference. The measured region contains no synchronization between CPU and GPU, so it measured only host latency. 

The ncclAllReduce time is still concerning, but likely second-order. I suspect, due to some CPU activity and `tokens_per_expert.cpu()`, ncclAllReduce kernels across ranks didn't start at the same time but had to finish at the same time. So, the one that started earlier may appeared to run longer. Is this explanation plausible? @nsarka and @cowanmeg

**Comment by wujingyue** [comment]
I'm closing this for now. When we switch from a for loop to grouped gemms, the host latency impact will be minimized and we'll likely see a very different timing.
